contracts/adapters/IPRebalanceExtension.sol

```diff
@@ -0,0 +1,478 @@
+        uint256 currentRawUnderlying = setToken.getDefaultPositionRealUnit(transformInfo.underlyingComponent).toUint256();
+        uint256 maxTransformUnits = transformInfo.maxTransformSize.preciseDiv(setToken.totalSupply());
+
+        if (_transformRemaining) {
+            require(maxTransformUnits >= currentRawUnderlying, "transform units greater than max");
+            unitsToTransform = currentRawUnderlying;
+        } else {
+            uint256 currentUnits = setToken.getDefaultPositionRealUnit(_transformComponent).toUint256();
+            uint256 exchangeRate = transformInfo.transformHelper.getExchangeRate(transformInfo.underlyingComponent, _transformComponent);
+
+            uint256 currentUnitsUnderlying = currentUnits.preciseDiv(exchangeRate);
+            uint256 targetUnitsUnderlying = rebalanceParams[_transformComponent];
+
+            unitsToTransform = targetUnitsUnderlying > currentUnitsUnderlying ? targetUnitsUnderlying.sub(currentUnitsUnderlying) : 0;
+
+            if (unitsToTransform > currentRawUnderlying) {
+                unitsToTransform = currentRawUnderlying;
+            }
+        }
+
+        if (unitsToTransform > maxTransformUnits) {
+            unitsToTransform = maxTransformUnits;
+        }
+
+
+        require(unitsToTransform > 0, "nothing to transform");
+
+        (address module, bytes memory callData) = transformInfo.transformHelper.getTransformCall(
+            manager.setToken(),
+            transformInfo.underlyingComponent,
+            _transformComponent,
+            unitsToTransform,
+            _transformData
+        );
+
+        invokeManager(module, callData);
+    }
+
+    /**
+     * Parameterizes the rebalancing trades through GeneralIndexModule. For all transform components, the target units
+     * remain fixed at their current allocations. For non-transform units, the target units are calculated by taking into
+     * account the total amount of the component needed to create the target trasnform units.
+     */
+    function _startGIMRebalance() internal {
+        
+        uint256[] memory rebalanceTargets = new uint256[](setComponentList.length);
+
+        for (uint256 i = 0; i < setComponentList.length; i++) {
+
+            address component = setComponentList[i];
+
+            if (_isTransformComponent(component)) {
+                rebalanceTargets[i] = setToken.getDefaultPositionRealUnit(component).toUint256();
+            } else {
+
+                uint256 finalTotalUnderlyingUnits = _getFinalTotalUnderlyingUnits(component, setComponentList);
+                uint256 currentTotalUnderlyingUnits = _getCurrentTotalUnderlyingUnits(component, setComponentList);
+
+                uint256 targetUnderlying = rebalanceParams[component];
+                int256 diff = finalTotalUnderlyingUnits.toInt256() - currentTotalUnderlyingUnits.toInt256() + targetUnderlying.toInt256();
+
+                rebalanceTargets[i] = diff > 0 ? diff.toUint256() : 0;
```

Comment:
What exactly is the `diff` param? Is this the amount of underlying tokens we expect to be left over after all the transforms are done at the end of the rebalance?

Reply:
It is really just an intermediate value. The correct rebalance target is equal to max(diff, 0). I've added a lot more detail to the javadoc to help understand what is going on here.